### PR TITLE
Making rchk happier, see #845

### LINKED ISSUE #845: Make `rchk` happy
Currently getting 
````
ERROR: too many states (abstraction error?) in function strptime_internal

Function DataFrameScanFunction::dataframe_scan_bind(duckdb::ClientContext&, std::__1::vector<duckdb::Value, std::__1::allocator<duckdb::Value> >&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, duckdb::Value, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, duckdb::Value> > >&, std::__1::vector<duckdb::LogicalType, std::__1::allocator<duckdb::LogicalType> >&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >&)
  [UP] ignoring variable df as it has address taken, results will be incomplete 

Function cpp_str_to_strsexp(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >)
  [UP] unprotected variable retsexp while calling allocating function cpp_str_to_charsexp(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:144
  [PB] has negative depth /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:145
  [UP] attempt to unprotect more items (1) than protected (0), results will be incomplete /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:145
  [PB] has possible protection stack imbalance /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:147

Function cstr_to_charsexp(char const*)
  [PB] has possible protection stack imbalance /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:130

Function duckdb_execute_R
  [UP] unprotected variable retlist while calling allocating function Rf_allocVector /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:410
  [UP] unprotected variable retlist while calling allocating function Rf_allocVector /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:415
  [UP] unprotected variable retlist while calling allocating function Rf_allocVector /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:425
  [UP] unprotected variable retlist while calling allocating function Rf_allocVector /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:428
  [UP] unprotected variable retlist while calling allocating function Rf_allocVector /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:484
  [UP] unprotected variable retlist while calling allocating function Rf_mkChar /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:485
  [UP] unprotected variable retlist while calling allocating function Rf_mkChar /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:486
  [UP] unprotected variable retlist while calling allocating function Rf_setAttrib(?,S:class,V) /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:487
  [UP] unprotected variable retlist while calling allocating function Rf_install /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:488
  [UP] unprotected variable retlist while calling allocating function Rf_mkString /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:488
  [UP] unprotected variable retlist while calling allocating function Rf_setAttrib(?,S:tzone,V) /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:488
  [UP] unprotected variable retlist while calling allocating function Rf_mkString /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:502
  [UP] unprotected variable retlist while calling allocating function Rf_setAttrib(?,S:class,V) /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:502
  [UP] unprotected variable retlist while calling allocating function Rf_mkString /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:527
  [UP] unprotected variable retlist while calling allocating function Rf_setAttrib(?,S:class,V) /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:527
  [UP] unprotected variable retlist while calling allocating function Rf_install /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:528
  [UP] unprotected variable retlist while calling allocating function Rf_mkString /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:528
  [UP] unprotected variable retlist while calling allocating function Rf_setAttrib(?,S:units,V) /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:528
  [UP] unprotected variable retlist while calling allocating function Rf_mkCharCE /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:566
  [PB] has negative depth /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:581
  [UP] attempt to unprotect more items (1) than protected (0), results will be incomplete /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:581
  [PB] has possible protection stack imbalance /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:585

Function duckdb_prepare_R
  [UP] unprotected variable retlist while calling allocating function cpp_str_to_strsexp(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >) /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:233
  [PB] has negative depth /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:235
  [UP] attempt to unprotect more items (1) than protected (0), results will be incomplete /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:235
  [PB] has possible protection stack imbalance /opt/R-svn/packages/build/1UhIcztM/duckdb/src/duckdbr.cpp:289
Analyzed 23415 functions, traversed 201029 states.
````

### PR DESCRIPTION
**Pull request by hannes**
Some changes in PROTECT/UPROTECT

### CODE REVIEW

**Comment by hannesmuehleisen** [comment]
CC @pat-s & @krlmlr

File: tools/rpkg/src/duckdbr.cpp [hunk 1]
```diff
@@ -127,6 +127,7 @@ static SEXP cstr_to_charsexp(const char *s) {
 		Rf_error("cpp_str_to_charsexp: Memory allocation failed");
 		UNPROTECT(1);
```
Comment (krlmlr) [inline tools/rpkg/src/duckdbr.cpp:128]: `Rf_error()` makes a long jump, no need to `UNPROTECT()` this path.

File: tools/rpkg/src/duckdbr.cpp [hunk 2]
```diff
@@ -214,29 +215,27 @@ SEXP duckdb_prepare_R(SEXP connsexp, SEXP querysexp) {
 	auto stmtholder = new RStatement();
 	stmtholder->stmt = move(stmt);
 
-	SEXP stmtsexp = PROTECT(R_MakeExternalPtr(stmtholder, R_NilValue, R_NilValue));
-	R_RegisterCFinalizer(stmtsexp, (void (*)(SEXP))duckdb_finalize_statement_R);
-
 	SEXP retlist = PROTECT(NEW_LIST(6));
 	if (!retlist) {
-		UNPROTECT(2); // retlist, stmtsexp
+		UNPROTECT(1); // retlist
```
Comment (krlmlr) [inline tools/rpkg/src/duckdbr.cpp:220]: Can we make `NEW_*()` throw an error to simplify the code here and everywhere?

**Review by krlmlr — APPROVED** [review]

File: tools/rpkg/src/duckdbr.cpp (hunk 1 above)
Comment (hannesmuehleisen) [reply tools/rpkg/src/duckdbr.cpp:128]: that's a good point

File: tools/rpkg/src/duckdbr.cpp (hunk 2 above)
Comment (hannesmuehleisen) [reply tools/rpkg/src/duckdbr.cpp:220]: NEW_* are R API macros and we have no control over their behaviour

**Comment by hannesmuehleisen** [comment]
It also looks like vector allocation does not return `NULL`, ever: https://github.com/wch/r-source/blob/trunk/src/main/memory.c#L2826

**Pull request merged**